packages/server/src/game.ts

```diff
@@ -767,47 +783,61 @@ function updateUnit(dt: Milliseconds, g: Game, unit: Unit, presence: PresenceMap
+                    throw "[game] Unit ordered to build something it can't";
 
-            if (bp.buildCost > owner.resources) {
-                console.info("[game] Unit ordered to build but player doesn't have enough resources");
-                clearCurrentAction();
-                break;
-            }
+                if (buildCapability.buildCost > owner.resources)
+                    throw "[game] Unit ordered to build but player doesn't have enough resources";
 
-            const BUILDING_DISTANCE = 2;
-            switch(moveTowards(cmd.position, BUILDING_DISTANCE)) {
-            case 'Unreachable':
-                clearCurrentAction();
-                break;
-            case 'ReachedTarget':
-                owner.resources -= bp.buildCost;
-                // TODO - this should take time
-                // already specified in bp.buildTime
+                const buildingData = getUnitDataByName(cmd.building);
+                if (!buildingData)
+                    throw "[game] Unit ordered to build unknown unit" +  cmd.building;
 
-                g.lastUnitId += 1;
-                g.units.push(createUnit(
-                    g.lastUnitId,
-                    unit.owner,
-                    cmd.building,
-                    { x: cmd.position.x, y: cmd.position.y },
-                ));
+                const getBuildingComponentFromUnitData = (ud: UnitData) => {
+                    return ud.find(c => c.type === 'Building') as Building;
+                };
+
+                const buildingComponent = getBuildingComponentFromUnitData(buildingData);
+                if (!buildingComponent)
+                    throw "[game] Unit ordered to build something that's not a building: " + cmd.building;
+
+                if (!mapEmptyForBuilding(g.board.map, buildingComponent.size, cmd.position))
+                    throw "[game] Unit ordered to build but some tiles are obscured";
+
+                const BUILDING_DISTANCE = 2;
+                switch(moveTowards(cmd.position, BUILDING_DISTANCE)) {
+                case 'Unreachable':
+                    clearCurrentAction();
+                    break;
+                case 'ReachedTarget':
+                    owner.resources -= buildCapability.buildCost;
+                    // TODO - this should take time
+                    // already specified in bp.buildTime
+
+                    g.lastUnitId += 1;
+                    g.units.push(createUnit(
+                        g.lastUnitId,
+                        unit.owner,
+                        cmd.building,
+                        { x: cmd.position.x, y: cmd.position.y },
+                    ));
+                    clearCurrentAction();
+                    break;
+                }
+                break;
+            }
+            catch(e) {
+                console.info(e);
                 clearCurrentAction();
                 break;
             }
-            break;
+
```

Comment:
I don't really like that everything hinges on the break here, i think those blocks are too long but maybe refactoring them is out of scope for this PR